File: Longest Prefix Suffix.cpp
class Solution {
  public:
    int lps(string str) {
        long long pre = 0, suf = 0, base = 29, p = 1, ind = 0, n = str.size(), mod = 1e9+7;
        
        for(int i=0; i<n-1; i++){
            int cha = str[i] - 'a' + 1;
            int chp = str[n-i-1] - 'a' + 1;
            
            pre = ((pre*base)%mod + cha)%mod;
            suf = (suf + (chp*p)%mod)%mod;
      
            p=(p*base)%mod;
            if(pre == suf){
                ind = i+1;
            }
        }
        return ind;
    }
};

File: Longest Prefix Suffix.java
class Solution {
    int longestPrefixSuffix(String s) {
        int n = s.length();
        int[] lps = new int[n];
        int j = 0; 

        for (int i = 1; i < n; i++) {
            
            while (j > 0 && s.charAt(i) != s.charAt(j)) {
                j = lps[j - 1]; 
            }
            if (s.charAt(i) == s.charAt(j)) {
                j++;
            }
            lps[i] = j;
        }

        
        return lps[n - 1] == n ? 0 : lps[n - 1];
    }
}
